### DISCUSSION TITLE
Deployment patterns

### DISCUSSION THREAD

**Discussion by leighmcculloch** [stellar-deprecated/starlight#159] [Discussion]
@nikhilsaraf posed the question to me recently about how payment channels might be deployed, and if we would have some sort of service that a user could deploy and use to watch the network for bad actors closing channels in old states.

I think this also raises the question of what components will a user of a payment channel likely have, or how might they want to organize monitoring, etc. Regardless of whether we build these components, or only build the building blocks for them, I think it is worthwhile we hash this out a little in parallel with the work we're doing.

There could probably be many types of deployments, so the goal of this discussion should be to capture different possible deployments so we can develop and understanding of how the SDK might be used.

cc @nikhilsaraf @acharb @accordeiro

**Comment by leighmcculloch** [comment]
I think there could be a deployment that separates the responsibilities of proposing payments, agreeing to proposals, and monitoring for safety, into three components:

1. A `watcher` that monitors the state of the escrow accounts on the network, has a copy of the latest agreement, and in the event an older agreement is published to the network, it submits the latest declaration transaction to ensure the latest agreement is preserved. This would be a long running service.

2. A `server` that receives new agreements from the other participant and signs them if the new agreements result in the participant gaining funds, and the new agreement is valid and won't break the channel. The server also shares the latest agreement with the `watcher` so it knows about that new state. This would be a long runner service.

3. A `client` that connects to the other participant's `server` to make payments. The client also shares the latest agreement with the `watcher` so it knows about the new state. This could be short-lived logic embedded in an application.

In diagram form:

![Untitled Diagram](https://user-images.githubusercontent.com/351529/125342448-ca1c5f80-e309-11eb-9f10-4f778d21d3a5.png)

[Diagram source](https://viewer.diagrams.net/?highlight=0000ff&edit=_blank&layers=1&nav=1#R7Vtbk6I4FP41PtoFJFx8HJ3enoeZmqmytuY5QhqpRWJBbNv%2B9RskQZPgrCtB44wvFpxACOec7zuX4AjMVu8vJVovv5EE5yPPSd5H4PPI88Jwwn5rwa4RuM7EaSRpmSVcdhDMsw8sLuTSTZbgSrqQEpLTbC0LY1IUOKaSDJUl2cqXvZJcfuoapVgTzGOU69KfWUKXjTTynYP8C87SpXiy6%2FCRFRIXc0G1RAnZHonA8wjMSkJoc7R6n%2BG8Vp7QS3PfXydG24WVuKDn3OA1N7yhfMPf7Sei8RKXfHl0J965JJsiwfVt7ghMt8uM4vkaxfXollmZyZZ0lfNhfRl8ZW%2B4pPj9SMSX9YLJCtNyxy4Ro0Jn3EkA5Ofbg8ZdocblkbYDLkPcyGk79UEP7ICrolstQFPLLM%2Fqd7m5VqCsFd%2B5olagrVqBkxtqJbAWQj68IYRCTS2aPnCRfKrZmJ3FOaqqLD5TBRXZlDGWoEpRmWIqkRpOJBbX1XSkBr9DC0JW4hzR7E3m%2Fi7V8Cf8IFmNCWEF4MtWaJ1VTNG8Db%2FrmKuViaDCiBAqEzU60CbaW6p97bOM5%2Fqa9f5eJ4jivdaY0rwgZ1qaLpiTB2l9xFSEq3qxKC0xXnWxQm2Pr2jBcgHJyijP0qJ2AXYPAw2Y1g6esWD7iQ%2BssiSp55iWuMo%2B0GI%2Fn8PO1%2FWb7t%2Fdn478z50eI9xQBU2bHPD5RsfxtwtMY%2BfJ9SKvn0uIS8jra4X7Gim6EsKgjrDggbC%2BCNPDxp0iTLihEYRBx5MsMLYJcBPNZF9ImX2QQjMES%2BvX9WG8yzMW8kvw3%2FF%2B0SQHXxetAMX%2FpPuU4fuGsmkwl1dNOcQo2kiSoIIHBHqSEHRgNzKQI4h6bnAK83QKmzworC%2BF6SneN1JklJQ6eaE4Zp5M9YGK7jnPBiJrvdEEk3kOcGVcWURkrqtZbhjcBQ%2FcDYA7PfG7b9ydilkXZRCOVUDzrgQ00FEG24U0LzSENOBdD2l6xveCO8B0FygTrtgbZc6TAwGUbDDuCTph2lCaNVQMawaSem93IEhan3Sag2SkTOQOBklPLxnuF5LCFQ1A0vVDOd90jbjIWMmP1KrPCCQ7NhY6DTTnpzhfkO3zQTDdC9iAsE4t%2BjPKRmXjA6qd%2B7PTV3gZgpmO0e7oMu72pylHWTBwpP1JdtDMeLEr6Y3r%2BWaxyjqIQPTTNOawq73WgsNE3A4mkRy3vX5uaJYG9JaoHTRgfRXrKoG8rWr%2FdyBX4ammYIZoAKoLNk0Demvq3mlAgOP3p4Fz9rVuQQPWl9haNnApDWjZgDMMDQClaea7ZmkA6MXC9zUu6gdoeJ%2FlpLKkLjC7pcaSQknJPbEuZg5A16xmmUBvv9jBBB074g8m6JcQBAMzgb7jcRdMIBBgggn8cCIzgZmmnavkEmAAJhC9tON8DpdvFnxj5yoNkkDEnKt8pqr3Mi1RC4xuqZZz%2BkkDZYTCJI9AcDHYYRfYi2Rvjp09FVnrZSa4OfQcOZ8y074V8UOgcAhq1jtuw4CtI%2BkSYeEBtsvB1hVCLASb8DIjYIvCyIjBx4KqxO7IAPiCpjZHls23itRgQ6Qj%2Fln2ZfCpP4b0RqQKbUN1UKDkTb5ntg6Cpz%2FsV8ugX3%2BFbHm7FBoLzs6TE%2FlyGB33DM5m%2BcHUrskA%2FODp%2FGBXxHaVKuXS7VN1O2OgbROVHAw3SaC%2Ba%2FJ7koPJvRSbyIGdHv463Fx%2B%2BAM2eP4X)

**Reply by leighmcculloch** [reply]
I think something that becomes challenging in this model is synchronized access to the data store that the client and server use. A server and client must never sign different transactions for the same iteration. A high-throughput application could have significant contention on database locks.

**Comment by leighmcculloch** [comment]
I think there could be a deploy that embeds most responsibility into a client or daemon, although where something external can monitor for bad behavior as well. I think this is more similar to what @nikhilsaraf found elsewhere in research. 

1. A `watcher` that monitors the state of the escrow accounts on the network, has a copy of the latest agreement, and in the event an older agreement is published to the network, it submits the latest declaration transaction to ensure the latest agreement is preserved. This would be a long running service.

2. A `client` that opens channels, proposes new states for the channel, and agrees to new states proposed by the other participant. This would be a long running service, always-on and persistently connected to the other participant.

In diagram form:

![Untitled Diagram](https://user-images.githubusercontent.com/351529/125357736-df4eb980-e31c-11eb-911c-2f7f9558fcf6.png)

[Diagram source](https://viewer.diagrams.net/?highlight=0000ff&edit=_blank&layers=1&nav=1#R7VvRcuI2FP0aZtqHMLYlGXgMSTb7sNvdKdPZbN%2BErWC1xmZsEWC%2FvhKWjCU5WwbkYNK%2BMPaVLEvn3nOke4EBuFtuHwu8Sj7nMUkHgRdvB%2BB%2BEAQTNOafwrCrDGMEK8OioHFl8g%2BGGf1BpNGT1jWNSal1ZHmeMrrSjVGeZSRimg0XRb7Ruz3nqf7WFV4QyzCLcGpbv9GYJWoV3sH%2BkdBFot7se7JliVVnaSgTHOebhgk8DMBdkeesulpu70gqsFO4VM99eKW1nlhBMnbMA0H1wAtO13Jt3zCLElLI6bGdWnORr7OYiMf8AZhuEsrIbIUj0brhTua2hC1T2WxPQ87shRSMbBsmOa1Hki8JK3a8i2pVmMkYAVDebw6I%2BwrGpIF2KG1YOnlRD33AgV9IKNphARYsdykVa7k4KlBHBXlviArsKypwckFUwt5SCMELUmhiwfIxL%2BiPPLNg4Rq4EpfRLqUcnwL8OzjzCslP89qAo78Xe3y%2FrBkfhkh7We0dPnKDKEAGoqGNaNgC6NgBoGrzayBqQUmy%2BFbsbwLNFJcljY4MqjJfFxHR9gSGiwVhmjNJrO2LNkwNGFALDMpWkBQz%2BqLvpm3YyDd8zalQmde8UNNfDVGtRj7V3P2MgaCxx0BoDFRhYA20d1W97OO8N7K89znPKMu5TIQpEyEsrhbiCkcRj2RmN5QMM2I7nXvlE57zI5bma5zSRSYCgfuaixGYijCn%2FBBzKxuWNI7FGNOCcJrg%2BX48j9%2BvxHr3CKDpAN23x42KRpM79alLDjhoHmzaOHXjDQMP%2BDqvzosT1SV%2Ffi7J2Z7zLc91w7vwf951wLvxO%2BPda3vWKbzzvV4RzU5GuiEa7DvRUOAPERjD0K8%2BocaWEJxIOxC8He3s498jaWHWVVBOxeXZlPOGHgS6M2%2FOZGCdLWujqim75aedFXfDT9B3fvL5DBU3xScCjgg6Ngjqd0bQwM4mrpegKjAdENRHI%2F0o6juJmJtAp70fdkDQlgJNq4Nm8pak83zzcDBM9wbeoLwjTP%2BNjNIoIEGzAnL0yRaexmCOMd41usmwf12BjAkDT6vz8otqxJNDCVmhNFvPl7RFCDjwpGxTjkVByLKtPHgRkajJ4WIXDydjfRcPzgtDtzJgVyT7IQO9T3D9kcEqdOpGbtJz1I0MQHPCrmXArlpduwwocrx%2FGbArH53LAIe52D0Jdwx9iJThe7P1fiu9Vd0J5G7gENSdv5KC8lUKz%2B%2B7HaUsvc8YrAPGqcpiHTC8bpQFGCU65LtVFmDnH19WJBMvsCTkLs3LnqQaNamclN%2F4OVMD%2BUz5UCOHoG1Ut%2BJi13feUlzG6A3FpfflwqsTlzq96kpc7K9srkJcFKlciAsaTXRxcVRoNE48oANxga7qGEn1iwN2nLxsKXtqXH%2FfK80IyduDtIib3ZkCEvZKQEJjsw%2BdlT9MJXIkIKFRKEWBWwGBdvnjj1W8L3ia%2BrGh3EHXmg3VTHORDY3RRNeGM0unbjXFVVHkJE3xnGtKS8YT9EpTfORIU8wSR0elFFNRHB9JoF1JeZ%2BK4rK%2B0mtFyf6My8cPBX68nf3218Pvxe32KblpcbPtlzbRKHkYM%2Ffnk0PO1MyX1GOvpEwHETK05BRVgpcUIWj84gGdnBkZKoSOzIycxZVdzHvYRgnOuH%2F4JJQGlLY%2B%2FLIiRUlLtlcIxUGaZ79eTi%2BO1gY93n5CuJ%2BJiBcYInJmleUcEeG3h%2F92VN0Pf5ABD%2F8A)

**Reply by nikhilsaraf** [reply]
I think having the `watcher` node/service is definitely going to play a key part. I can also see the community building this into a SaaS for anyone to sign up to be their watcher node for redundant protection.

**Comment by nikhilsaraf** [comment]
As a third option, also want to include the "mailbox" component in our list that we had discussed, which is slightly different from the watcher/client/server model. This watcher/client/mailbox pattern isn't fully applicable today with our high-throughput use-case but an important component to list since we're listing components here.

Both server and mailbox are long running services that can serve to operate as proxies for the client when client is offline, but the key difference between the server and mailbox is that the client always has to go through the mailbox to receive messages (which also gets around the synchronization in the client/server model) and the mailbox will never auto-approve payments even if the client will gain funds.

I'd like to analogize the mailbox approach to email, whereas our fast throughput approach is like IM. We are focusing on the IM case for now and when we get to the email case (low throughput) we can consider the mailbox component.

**Reply by leighmcculloch** [reply]
I like the mailbox approach as well, but CAP-21 requires a multi-step signing process on each agreement which makes mailbox problematic. For example, I can send you a payment, and you can accept it, but it isn't authorized until I send you a second follow up message finalizing it. To do a single message in either direction, like email, we'd need another change to the Stellar protocol in addition or instead of CAP-21. I've been discussing this specific limitation with @stanford-scs [in Slack](https://stellarfoundation.slack.com/archives/C0166LZF4VC/p1626130667124200). I don't have a good idea yet for how we could change the protocol to make this work better like email.

**Reply by leighmcculloch** [reply]
I think we can make this deployment pattern possible if we implement this proposal: https://github.com/stellar/experimental-payment-channels/issues/165.

**Comment by shanzzam** [comment]
@leighmcculloch and @nikhilsaraf: now that CAP-40 is published does this mean you want to move forward designing the "mailbox" as part of the model as outlined above? 

@nikhilsaraf
> "This watcher/client/mailbox pattern isn't fully applicable today with our high-throughput use-case"
Can you clarify this? Do you mean this is a component we need for future use-cases but we should probably build now?

@leighmcculloch the diagram is very useful, thank you for taking the time to create that.
